### PR TITLE
[DUOS-2329][risk=no] New automation test framework

### PR DESCRIPTION
**Pull request by rushtong**
## Addresses
https://broadworkbench.atlassian.net/browse/DUOS-2329

This PR adds infrastructure for a new automation/performance/smoke-testing framework. The existing one in `/automation` is bound to a jenkins job which has significant drawbacks. This approach sets us up for running these tests in a GHA-created BEE and subsequently reporting the results back to QA.

Future stories and PRs will address running these tests in a BEE, reporting tests back to QA, and then deprecating/removing everything in `/automation`.

----
Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

- Label PR with a Jira ticket number and include a link to the ticket
- Label PR with a security risk modifier [no, low, medium, high]
- PR describes scope of changes
- Get a minimum of one thumbs worth of review, preferably two if enough team members are available
- Get PO sign-off for all non-trivial UI or workflow changes
- Verify all tests go green
- Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/test/java/org/broadinstitute/dsp/ontology/performance/scenarios/Matching.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        "purpose": {
+          "type": "named",
+          "name": "http://purl.obolibrary.org/obo/DOID_162"
+        },
+        "consent": {
+          "type": "named",
+          "name": "http://purl.obolibrary.org/obo/DOID_162"
+        }
+      }
+            """;
```
Comment (connorlbark) [inline src/test/java/org/broadinstitute/dsp/ontology/performance/scenarios/Matching.java]: nit: format

File: src/test/java/org/broadinstitute/dsp/ontology/performance/TestRunner.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+  private final TestConfig config = new TestConfig();
+
+  private final HttpProtocolBuilder protocol = http
+      .baseUrl(config.getBaseUrl())
+      .header("X-App-ID", "DUOS")
+      .doNotTrackHeader("1")
+      .userAgentHeader("Mozilla/5.0 (Windows NT 5.1; rv:31.0) Gecko/20100101 Firefox/31.0");
+
+  {
```
Comment (connorlbark) [inline src/test/java/org/broadinstitute/dsp/ontology/performance/TestRunner.java:35]: I'm confused at the syntax here - it's just a block on its own?

**Review by connorlbark — COMMENTED** [review]
Changes look great, but locally my tests failed:
image

It looks like something crashed ontology.

```
ontology | ERROR [2023-02-13 16:22:58,169] io.dropwizard.jersey.errors.LoggingExceptionMapper: Error handling a request: 5e6eeb07839b8b34
ontology | ! java.net.ConnectException: Timeout connecting to [elasticsearch5a1.dsde-dev.broadinstitute.org/34.123.139.204:9200]
ontology | ! at org.apache.http.nio.pool.RouteSpecificPool.timeout(RouteSpecificPool.java:169)
ontology | ! at org.apache.http.nio.pool.AbstractNIOConnPool.requestTimeout(AbstractNIOConnPool.java:632)
ontology | ! at org.apache.http.nio.pool.AbstractNIOConnPool$InternalSessionRequestCallback.timeout(AbstractNIOConnPool.java:898)
ontology | ! at org.apache.http.impl.nio.reactor.SessionRequestImpl.timeout(SessionRequestImpl.java:198)
ontology | ! at org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor.processTimeouts(DefaultConnectingIOReactor.java:213)
ontology | ! at org.apache.http.impl.nio.reactor.DefaultConnectingIOReactor.processEvents(DefaultConnectingIOReactor.java:158)
ontology | ! at org.apache.http.impl.nio.reactor.AbstractMultiworkerIOReactor.execute(AbstractMultiworkerIOReactor.java:351)
ontology | ! at org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager.execute(PoolingNHttpClientConnectionManager.java:221)
ontology | ! at org.apache.http.impl.nio.client.CloseableHttpAsyncClientBase$1.run(CloseableHttpAsyncClientBase.java:64)
ontology | ! ... 1 common frames omitted
ontology | ! Causing: java.net.ConnectException: Timeout connecting to [elasticsearch5a1.dsde-dev.broadinstitute.org/34.123.139.204:9200]
ontology | ! at org.elasticsearch.client.RestClient.extractAndWrapCause(RestClient.java:930)
ontology | ! at org.elasticsearch.client.RestClient.performRequest(RestClient.java:300)
ontology | ! at org.elasticsearch.client.RestClient.performRequest(RestClient.java:288)
ontology | ! at org.broadinstitute.dsde.consent.ontology.service.ElasticSearchSupport.lambda$retryRequest$2(ElasticSearchSupport.java:136)
ontology | ! at com.github.rholder.retry.AttemptTimeLimiters$NoAttemptTimeLimit.call(AttemptTimeLimiters.java:78)
ontology | ! at com.github.rholder.retry.Retryer.call(Retryer.java:160)
ontology | ! ... 67 common frames omitted
ontology | ! Causing: com.github.rholder.retry.RetryException: Retrying failed to complete successfully after 3 attempts.
ontology | ! at com.github.rholder.retry.Retryer.call(Retryer.java:174)
ontology | ! at org.broadinstitute.dsde.consent.ontology.service.ElasticSearchSupport.retryRequest(ElasticSearchSupport.java:143)
ontology | ! ... 66 common frames omitted
ontology | ! Causing: javax.ws.rs.InternalServerErrorException: HTTP 500 Internal Server Error
ontology | ! at org.broadinstitute.dsde.consent.ontology.service.ElasticSearchSupport.retryRequest(ElasticSearchSupport.java:156)
ontology | ! at org.broadinstitute.dsde.consent.ontology.service.ElasticSearchAutocomplete.executeSearch(ElasticSearchAutocomplete.java:76)
ontology | ! at org.broadinstitute.dsde.consent.ontology.service.ElasticSearchAutocomplete.lookup(ElasticSearchAutocomplete.java:140)
ontology | ! at org.broadinstitute.dsde.consent.ontology.service.ElasticSearchAutocomplete.lookup(ElasticSearchAutocomplete.java:135)
ontology | ! at org.broadinstitute.dsde.consent.ontology.resources.AutocompleteResource.getTerms(AutocompleteResource.java:32)
ontology | ! at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
ontology | ! at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
ontology | ! at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
ontology | ! at java.base/java.lang.reflect.Method.invoke(Unknown Source)
ontology | ! at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
ontology | ! at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:134)
ontology | ! at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:177)
ontology | ! at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:219)
ontology | ! at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:81)
ontology | ! at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:478)
ontology | ! at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:400)
ontology | ! at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81)
ontology | ! at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:255)
ontology | ! at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
ontology | ! at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
ontology | ! at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
ontology | ! at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
ontology | ! at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
ontology | ! at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
ontology | ! at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234)
ontology | ! at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:684)
ontology | ! at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394)
ontology | ! at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346)
ontology | ! at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:358)
ontology | ! at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:311)
ontology | ! at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205)
ontology | ! at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
ontology | ! at org.eclipse.jetty.servlet.ServletHandler$ChainEnd.doFilter(ServletHandler.java:1656)
ontology | ! at io.dropwizard.servlets.ThreadNameFilter.doFilter(ThreadNameFilter.java:35)
ontology | ! at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
ontology | ! at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626)
ontology | ! at io.dropwizard.jersey.filter.AllowedMethodsFilter.handle(AllowedMethodsFilter.java:47)
ontology | ! at io.dropwizard.jersey.filter.AllowedMethodsFilter.doFilter(AllowedMethodsFilter.java:41)
ontology | ! at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:193)
ontology | ! at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626)
ontology | ! at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:552)
ontology | ! at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
ontology | ! at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1440)
ontology | ! at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
ontology | ! at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:505)
ontology | ! at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
ontology | ! at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1355)
ontology | ! at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
ontology | ! at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
ontology | ! at com.codahale.metrics.jetty9.InstrumentedHandler.handle(InstrumentedHandler.java:284)
ontology | ! at io.dropwizard.jetty.RoutingHandler.handle(RoutingHandler.java:52)
ontology | ! at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:772)
ontology | ! at io.dropwizard.jetty.ZipExceptionHandlingGzipHandler.handle(ZipExceptionHandlingGzipHandler.java:26)
ontology | ! at org.eclipse.jetty.server.handler.RequestLogHandler.handle(RequestLogHandler.java:54)
ontology | ! at org.eclipse.jetty.server.handler.StatisticsHandler.handle(StatisticsHandler.java:181)
ontology | ! at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
ontology | ! at org.eclipse.jetty.server.Server.handle(Server.java:516)
ontology | ! at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:487)
ontology | ! at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:732)
ontology | ! at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:479)
ontology | ! at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
ontology | ! at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
ontology | ! at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
ontology | ! at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
ontology | ! at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
ontology | ! at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
ontology | ! at java.base/java.lang.Thread.run(Unknown Source)
```

Looks related to elastic search - so it may be that my local instance can't connect to it, and this is an expected error. Is that true? If so, then these changes all look good.

As an aside, I have really bad performance issues locally on ontology / consent, so I'm not surprised that my time ranges are all bad. I'm not worried about that for what that implies for production.

**Comment by rushtong** [comment]
> Looks related to elastic search - so it may be that my local instance can't connect to it, and this is an expected error. Is that true? If so, then these changes all look good.

ES is only visible on the VPN - that could be it

File: src/test/java/org/broadinstitute/dsp/ontology/performance/TestRunner.java (hunk 2 above)
Comment (rushtong) [reply src/test/java/org/broadinstitute/dsp/ontology/performance/TestRunner.java:35]: That's exactly right, see the quickstart for an example of the same syntax: https://gatling.io/docs/gatling/tutorials/quickstart/

**Comment by connorlbark** [comment]
> ES is only visible on the VPN - that could be it

That's definitely it. All of them were due to that - so if it works on your machine, then I'm happy to approve.

**Review by connorlbark — APPROVED** [review]
👍 only thought - do we want to delete old scala tests? or keep them for now?

File: src/test/java/org/broadinstitute/dsp/ontology/performance/TestRunner.java (hunk 2 above)
Comment (connorlbark) [reply src/test/java/org/broadinstitute/dsp/ontology/performance/TestRunner.java:35]: https://stackoverflow.com/questions/2943556/static-block-in-java
Huh. TIL. Very strange.

**Comment by rushtong** [comment]
> 👍 only thought - do we want to delete old scala tests? or keep them for now?

@connorlbark - we can't delete them yet, they're still in use in the older jenkins based test process. The goal here is to eventually migrate all of that to the action in this PR. It's a step-wise process however as some significant pieces (like jenkins QA test reporting to BQ) have not been written yet.

File: .github/workflows/performance.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      - uses: actions/setup-java@v3
+        with:
+          distribution: 'temurin'
+          java-version: 17
+      - uses: actions/cache@v3
+        with:
+          path: ~/.m2/repository
+          key: ${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}
+          restore-keys: |
+            ${{ runner.os }}-maven-
```
Comment (fboulnois) [inline .github/workflows/performance.yaml]:
```suggestion
      - uses: actions/setup-java@v3
        with:
          distribution: 'temurin'
          java-version: 17
          cache: 'maven'
```

`setup-java` has a built in maven cache, so I think you can do this.

**Review by fboulnois — APPROVED** [review]
One suggestion:

**Pull request merged**